### PR TITLE
Add feature_engineering module and test notebook

### PR DESCRIPTION
**Pull request by tm-jc-nacpil**
This PR adds the `feature_engineering.py` module to `povertymapping`, alongside a test notebook that demonstrates its use.
- `generate_features`: Generates the base features for an AOI based on DHS, OSM, Ookla, and VIIRS (nighttime lights) parameters. By default it uses StandardScaler but can use other sklearn scalers.
- `generate_labels`: Generates the label data for an AOI given a column. By default, it uses StandardScaler.

Addresses #142 

# Example Usage

```
# Make the AOI for DHS
country_osm = "east-timor"
ookla_year = 2019
nightlights_year = 2016
col_rename_config = 'tl'
dhs_household_dta_path = settings.DATA_DIR/"dhs/tl/TLHR71DT/TLHR71FL.DTA"
dhs_geographic_shp_path = settings.DATA_DIR/"dhs/tl/TLGE71FL/TLGE71FL.shp"

dhs_gdf = generate_dhs_cluster_level_data(
    dhs_household_dta_path, 
    dhs_geographic_shp_path, 
    col_rename_config="tl",
    convert_geoms_to_bbox=True,
    bbox_size_km=2.4
).reset_index(drop=True)
country_data = dhs_gdf.copy()

# Generate the features for the AOI. scaled_only limits the returned features to scaled features only, while features_only returns only features and excludes the columns of the passed AOI.
features = generate_features(dhs_gdf, country_osm, ookla_year, nightlights_year,col_rename_config, scaled_only=True, features_only=True)

# Generate the labels by scaling the wealth index column
labels = generate_labels(labels, 'Wealth Index')

<model training code here>

```

### CODE REVIEW

**Comment by butchtm** [comment]
> This PR adds the `feature_engineering.py` module to `povertymapping`, alongside a test notebook that demonstrates its use.
> 
> * `generate_features`: Generates the base features for an AOI based on DHS, OSM, Ookla, and VIIRS (nighttime lights) parameters. By default it uses StandardScaler but can use other sklearn scalers.
> * `generate_labels`: Generates the label data for an AOI given a column. By default, it uses StandardScaler.
> 
> # Example Usage
> ```
> # Make the AOI for DHS
> country_osm = "east-timor"
> ookla_year = 2019
> nightlights_year = 2016
> col_rename_config = 'tl'
> dhs_household_dta_path = settings.DATA_DIR/"dhs/tl/TLHR71DT/TLHR71FL.DTA"
> dhs_geographic_shp_path = settings.DATA_DIR/"dhs/tl/TLGE71FL/TLGE71FL.shp"
> 
> dhs_gdf = generate_dhs_cluster_level_data(
>     dhs_household_dta_path, 
>     dhs_geographic_shp_path, 
>     col_rename_config="tl",
>     convert_geoms_to_bbox=True,
>     bbox_size_km=2.4
> ).reset_index(drop=True)
> country_data = dhs_gdf.copy()
> 
> # Generate the features for the AOI. scaled_only limits the returned features to scaled features only, while features_only returns only features and excludes the columns of the passed AOI.
> features = generate_features(dhs_gdf, country_osm, ookla_year, nightlights_year,col_rename_config, scaled_only=True, features_only=True)
> 
> # Generate the labels by scaling the wealth index column
> labels = generate_labels(labels, 'Wealth Index')
> 
> <model training code here>
> ```

**Comment by butchtm** [comment]
Mistakenly closed instead of finishing my review. Reopening.

**Comment by butchtm** [comment]
Please check the parameters to the call to `generate_features` as it should not include the parameter `col_rename_config`  which is a parameter for generate dhs cluster data.

**Review by butchtm — CHANGES_REQUESTED** [review]

File: povertymapping/feature_engineering.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    # Add in OSM features
+    aoi = osm.add_osm_poi_features(aoi, country_osm, osm_data_manager)
+    aoi = osm.add_osm_road_features(aoi, country_osm, osm_data_manager)
+
+    # Add in Ookla features
+    aoi = ookla.add_ookla_features(aoi, "fixed", ookla_year, ookla_data_manager)
+    aoi = ookla.add_ookla_features(aoi, "mobile", ookla_year, ookla_data_manager)
+
+    # Add in the nighttime lights features
+    aoi = nightlights.generate_nightlights_feature(aoi, nightlights_year)
```
[the remark this answers is not in the document]
Comment (alronlam) [reply povertymapping/feature_engineering.py:66]: +1 to all the cache_dir updates to make it consistent to use `~/.geowrangler` already. Just noting you might have to move your cache files @tm-jc-nacpil  because of this.

File: povertymapping/feature_engineering.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    sklearn_scaler: Any = StandardScaler,
+    labels_only: bool = False,
+) -> pd.DataFrame:
+    """Generates labels for an AOI based on a specified column.
+
+    Args:
+        aoi (pd.DataFrame): The input AOI dataframe.
+        label_col (str): The column to use as the label.
+        inplace (bool, optional): Whether to overwrite the original dataframe or not. Defaults to False.
+        data_dir (str, optional): The path to the data directory. Defaults to settings.DATA_DIR.
```
Comment (alronlam) [inline povertymapping/feature_engineering.py:108]: This seems to be a wrong arg (it's not in the params)

File: povertymapping/feature_engineering.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        fill_na (bool, optional): Whether to fill missing values with fill_na_value or not. Defaults to True.
+        fill_na_value (int, optional): The value to fill missing values. Defaults to 0.
```
Comment (alronlam) [inline povertymapping/feature_engineering.py:110]: Please remove these too if you remove the params.

File: povertymapping/feature_engineering.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        aoi["label"] = scaler.fit_transform(aoi[["label"]])
+
+    if fill_na:
+        aoi = aoi.fillna(fill_na_value)
+
+    # Drop the input columns, leaving only the features
+    if labels_only:
+        aoi = aoi.drop(columns=input_cols)
+
+    return aoi
```
Comment (alronlam) [inline povertymapping/feature_engineering.py:140]: Actually, thinking about this more, the actual use-case we're doing is to just simply scale the label column right? Can we remove this function altogether? We might be over-engineering this. 

In usage without this function, it would just be as simple as something like this:
`gdf["Wealth Index"] = MinMaxScaler().fit_transform(gdf["Wealth Index"])`

**Review by alronlam — CHANGES_REQUESTED** [review]

**Review by alronlam — APPROVED** [review]

**Review by butchtm — APPROVED** [review]
